mixer/adapter/redisquota/redisquota_test.go

```diff
@@ -0,0 +1,865 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package redisquota provides a quota implementation with redis as backend.
+// The prerequisite is to have a redis server running.
+
+package redisquota
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/alicebob/miniredis"
+	"github.com/alicebob/miniredis/server"
+	"github.com/go-redis/redis"
+	"github.com/golang/glog"
+	lua "github.com/yuin/gopher-lua"
+	"istio.io/istio/mixer/adapter/redisquota/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/adapter/test"
+	"istio.io/istio/mixer/template/quota"
+)
+
+type (
+	LocalRedisServer struct {
+		redisServer     *server.Server
+		miniredisServer *miniredis.Miniredis
+		miniredisClient *redis.Client
+
+		scriptmap map[string]string
+	}
+
+	QuotaInfo struct {
+		algorithm    string
+		name         string
+		credit       int64
+		windowLength int64
+		bucketLength int64
+	}
+
+	RequestInfo struct {
+		bestEffort      bool
+		token           int64
+		timestamp       int64
+		allocated       int64
+		validDuration   int64
+		deduplicationid string
+	}
+)
+
+func NewLocalRedisServer(addr string) (*LocalRedisServer, error) {
+	mockServer := LocalRedisServer{
+		redisServer:     nil,
+		miniredisServer: nil,
+		miniredisClient: nil,
+		scriptmap:       nil,
+	}
+
+	var err error
+
+	mockServer.scriptmap = map[string]string{}
+
+	// initialize mini redis to handle redis command
+	mockServer.miniredisServer, err = miniredis.Run()
+	if err != nil {
+		panic(err)
+	}
+
+	mockServer.miniredisClient = redis.NewClient(&redis.Options{
+		Addr:     mockServer.miniredisServer.Addr(),
+		PoolSize: 10,
+	})
+
+	// initialize mock redis to handle EVAL
+	mockServer.redisServer, err = server.NewServer(addr)
+	if err != nil {
+		panic(err)
+	}
+
+	// register command handler to mock redis
+	mockServer.redisServer.Register("PING", func(c *server.Peer, cmd string, args []string) {
+		c.WriteInline("PONG")
+	})
+
+	runEvalFunc := func(c *server.Peer, script string, args []string) error {
+		L := lua.NewState()
```

Comment:
what is `args[0]` ? it is not used anywhere

Reply:
args[0] contains SHA1 digest of the script for EVALSHA command